### ISSUE TITLE
Multi materials issue

### ISSUE THREAD

**Issue by claro-lima** [nrgsille76/io_scene_max#7]
In models with more than one material, only the first one is applied to the entire mesh.

**Comment by nrgsille76** [comment]
@claro-lima  Can you upload the file wich makes problems? Or at least send me the console output? All error will be collected in console output.
Multi materials are usually supported, if they are assigned to the polygons and if they are located in the multi-/sub material directory. But if one material fails because of wrong data format then that part will be skipped.

**Comment by claro-lima** [comment]
Hi,
I think the problem is with Maxtree models. 
Other models from Dimensiva or 3dSky are imported correctly.
Great addon. Thank You.

**Comment by 3deutig** [comment]
hi together, we are a small cgi productionstudio and switching from max to blender about 6 Months ago and are very interested in such a tool and maybe could help you in some cases. for the problems from claro-lima. yes we could repruduze this also and found that in max/vray the Vray2SidedMtl is the problem, if this is pluged into the submaterial slot the import stops and the rest of the submaterial and material ids are gone. 

we found also another issue so if there are more than just one UV Channels only the first would be imported.

if you want we could easily prepares you scenes for it, please let me know.

please write me an mail so we can directly work on that. (jens@3deutig.com)

bye,
Jens

**Comment by nrgsille76** [comment]
@3deutig  Yea I removed importing mutiple UV maps because Blender does not support multiple UVs properly. Only the first imported is showing up, all other have vertices on zero position. The script now always checks the array compabilty of the UV map, and chooses the first map wich is compatible. I noticed that also boxmapping is included wich blender does not support.

By now only VrayMtl and VrayNormalmap is supported, blender always uses single sided faces for double sided faces they have to be doublicated and normal flipped wich I have not implemented in my script

**Comment by schroef** [comment]
> VrayMtl

If a material is double sided. Lots of them use translucency for that. So use a altered or dedicated map for this and add a translucency shader and mix it. Other addons also do this when other apps call it a "2 sided" material

**Comment by 3deutig** [comment]
hi everybody,
sorry for my late reply. 

as schroef wrote, blender supports multiple uv maps and on the fbx side it works. so there must be a way for importing this directly...hope so:,-)

for the vray specified thing, it is only a shader thing so can you please instead of stopp the import uses an alternative shader like you did for vraymtl?

anothe request ... it is possible to have an import to convert max layers to blender collections?

many thanks! and if we could help you with some sampe files please let us know.

bye,
jens

**Comment by SkipzNL** [comment]
Awesome plugin, I've got many older 3ds Max files I have direct access to now.
But I have the same issue as @claro-lima .

It's with XFrog models which have a Multi-Sub-Object material with Standard materials under them.
I can email you an .MAX scene if you'd like, I don't want to attach it here due to license restrictions.

**Comment by TLarchviz** [comment]
Hey there, really nice plugin!
Unfortunately got the same issue as others. And yes most likely 2 sided materials (so almost every proper made vegetation/tree/plant) are the culprit.
It's a pity because I don't really mind having to set up materials from scratch (i dont expect to have the addon import ready made materials) but if you loose ALL IDs + UV then it's unusable. And as previously mentioned it should be possible with blender since if you just import FBX you get all IDS + UV kept the same way they were in 3ds max.

Fix this and it's gonna be golden!

**Comment by nrgsille76** [comment]
@TLarchviz  Yes I am aware of this problem, multi materials are not supported with editable mesh objects. I have not found the chunk where the material information is stored. In an editable poly object the info is stored inside the face chunk together with the face id but this is not the case with an editable mesh object. Below I printed out the chunk ids, chunk length and the first 4 bytes inside the bracets.
![meshchunks](https://github.com/user-attachments/assets/88121b30-ea56-490b-a9d7-36fe08e297d2)

`0x906` unknown, includes only one number
`0x908` is mesh id
`0x914` vertex vectors
`0x912` faces vert ids
`0x924` flag, includes either 0 or 1
`0x928` flag, includes either 0 or 1
`0x92A` flag, includes either 0 or 1
`0x959` flag, includes either 0 or 1
`0x2398` uv key,0 for boxmap, 1 for texture coordinates
`0x2394` texture coordinates in UVW format
`0x2396` uv face ids must be the same amount as `0x912`

For material usually only one number per face is needed for the material slot but I do not know where this is stored in an editable mesh object. If anyone knows how to get that info, then I will be able to fix it.

**Comment by bicukow** [comment]
> @TLarchviz Yes I am aware of this problem, multi materials are not supported with editable mesh objects. I have not found the chunk where the material information is stored. In an editable poly object the info is stored inside the face chunk together with the face id but this is not the case with an editable mesh object. Below I printed out the chunk ids, chunk length and the first 4 bytes inside the bracets. ![meshchunks](https://private-user-images.githubusercontent.com/61376632/384895091-88121b30-ea56-490b-a9d7-36fe08e297d2.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ0Mzk5MjQsIm5iZiI6MTczNDQzOTYyNCwicGF0aCI6Ii82MTM3NjYzMi8zODQ4OTUwOTEtODgxMjFiMzAtZWE1Ni00OTBiLWE5ZDctMzZmZTA4ZTI5N2QyLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE3VDEyNDcwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYyMjgzZTZkOGY3ZGI3OTEzOWM0ODlhM2IyNWM4MzQ4NjBhY2E4MTY0M2I3YzgyMWQwMzUxMzljMjU5ZDQyMGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.EN8zIfp27CEmCyDbQhI0zWRH1Q69HOIFaTblV-2jPXA)
> 
> `0x906` unknown, includes only one number `0x908` is mesh id `0x914` vertex vectors `0x912` faces vert ids `0x924` flag, includes either 0 or 1 `0x928` flag, includes either 0 or 1 `0x92A` flag, includes either 0 or 1 `0x959` flag, includes either 0 or 1 `0x2398` uv key,0 for boxmap, 1 for texture coordinates `0x2394` texture coordinates in UVW format `0x2396` uv face ids must be the same amount as `0x912`
> 
> For material usually only one number per face is needed for the material slot but I do not know where this is stored in an editable mesh object. If anyone knows how to get that info, then I will be able to fix it.

how we can help you with that? I mean.. how you decoded the .max file to extract attributes?

**Comment by nrgsille76** [comment]
@bicukow  The missing data is in the editable mesh chunk, you can print the chunk type ( `hex(chunk.types)` ) and the data length ( `len(chunk.data)` The length of the face id chunk (0x0912) has to be the same as the uv chunk (0x0918 or 0x2396)
The definition can be found [here](https://github.com/nrgsille76/io_scene_max/blob/cb531c280d2b11701719c9c86c0fd862c53f0b35/source/import_max.py#L1577)

You can add `print("chunk", hex(chunk.types), len(chunk.data)` after line 1587 to print out the chunk type and data length. You can also use the function `get_float(chunk.data, count)` or `get_long(chunk.data, count)` to decode the chunk data.

**Comment by gntropy** [comment]
Situations where textures cannot be imported:  
1. Using CoronaBitmap instead of 3ds Max's built-in Bitmap to load images. It appears that the plugin cannot recognize CoronaBitmap.  
2. If the 3ds Max model is an Editable Mesh, the IDs are lost when imported into Blender, leaving only the first ID. Only Editable Poly can retain IDs. In 3ds Max itself, Editable Mesh can also store ID information, so it is unclear why the plugin cannot read it.  
For plant models, Editable Mesh is almost always used. In 3ds Max, collapsing to Editable Mesh provides smoother viewport performance compared to Editable Poly, making ID loss more likely to occur.  
We hope for improvements. Thank you.  

The above is the translation; apologies if the grammar is not perfect.

**Comment by Shmorrin** [comment]
> Hi, I think the problem is with Maxtree models. Other models from Dimensiva or 3dSky are imported correctly. Great addon. Thank You.

Hey. Could you please share how you import 3D models from 3dsky with materials on it? 
The only reason I still use 3ds Max (as an interior designer) is the ability to simply drag and drop models especially from 3dsky and have all the materials applied automatically. 
In Blender, meshes import fine with this addon, but the materials don’t come through correctly. I can see all the texture and material names listed in the Materials tab, but none of them actually appear in the shader or during rendering. The whole model ends up looking like plastic or chrome.
PBR textures are easy to set up manually, but the non-PBR textures that come with 99% of 3dsky models take forever to connect properly in Blender. Is there a quick or automated way to get these materials working?
I’d really appreciate any advice. Thanks!
